### DISCUSSION TITLE
why does a glob in single quotes on Windows not work?

### DISCUSSION THREAD

**Discussion by ashrasmun** [BurntSushi/ripgrep#2536] [Q&A] [answered]
#### What version of ripgrep are you using?

```
w:\sandbox\test>rg --version
ripgrep 13.0.0 (rev af6b6c543b)
-SIMD -AVX (compiled)
+SIMD +AVX (runtime)
```

#### How did you install ripgrep?

https://github.com/BurntSushi/ripgrep/releases

#### What operating system are you using ripgrep on?

Windows 10

#### Describe your bug.

I would like to search for text in a subset of files, not all of them. I tried using `-g/--glob`, but it doesn't seem to work.

#### What are the steps to reproduce the behavior?

```
w:\sandbox\test>type nul>test.txt

w:\sandbox\test>nvim test.txt # Add 'lib' to the file

w:\sandbox\test>rg lib
test.txt
lib

w:\sandbox\test>rg lib -g '*'
No files were searched, which means ripgrep probably applied a filter you didn't expect.
Running with --debug will show why files are being skipped.
```

#### What is the actual behavior?

https://gist.github.com/ashrasmun/245ec0e48dd2b8377b0aa4616b7e9b2d

#### What is the expected behavior?

I would like to search for string 'lib' in a narrow set of files, not all of them.

**Comment by BurntSushi** [comment, marked as answer]
Thanks for the report. ~Can you say why you didn't run with `--debug` and show the output, as the error message suggests?~

My best guess is that you have a Windows problem. IIRC, single quotes don't work the way your command suggests you think they do: https://superuser.com/questions/324278/how-to-make-windows-command-prompt-treat-single-quote-as-though-it-is-a-double-q

It might be a good idea to read a tutorial on how to use the Windows command line. I'm not a Windows user myself so I'm not sure what the best one might be.

**Comment by ltrzesniewski** [comment]
They actually provided the `--debug` output in the gist, and it says:

> `glob converted to regex: Glob { glob: "**/'*'", re: "(?-u)^(?:/?|.*/)'[^/]*'$"` [...]

So yeah, the single quotes are interpreted literally and they end up in the glob pattern.

**Comment by BurntSushi** [comment]
Ah nice! I completely missed the gist link. Thanks for pointing that out. :)

**Comment by ashrasmun** [comment]
Hey, thanks for the response. Everything works fine with double quotes. Sorry, I just rolled with what's written on https://github.com/BurntSushi/ripgrep/blob/master/GUIDE.md#manual-filtering-globs. Was there anything I could do better with gist to make it more visible? I tried to follow the guide when creating the issue as faithfully as I could.

**Reply by BurntSushi** [reply]
No you did great. I just respond to lots of issues and sometimes I miss stuff. I completely skipped right over the gist link.

And yeah, unfortunately there is an unresolveable conflict here between Windows and Linux/macOS, or at least, between a Windows shell and something like bash: single quotes are usually preferred in bash because they are interpreted literally. This is especially useful when special characters like `$` or `!` are used (the latter is used to negate a glob in ripgrep for example), since in bash, with double quotes, `$` and `!` have special significance. But on Windows, the single quote approach doesn't work. I don't really see a great way to resolve the issue in a non-clunky way. The docs generally assume basic proficiency with your shell environment. (And many of the docs are written with a shell like bash in mind, along with a standard POSIX user land.)
